src/test/java/com/diffmin/AppTest.java

```diff
@@ -216,6 +230,66 @@ private static void runTests(TestResources sources) throws Exception {
+                        private boolean isSetCollectionModified(CtElement element) {
+                            if (element.getRoleInParent() == CtRole.THROWN) {
+                                return ((CtExecutable<?>) element.getParent())
+                                        .getThrownTypes().stream()
+                                                .anyMatch(
+                                                        thrownType ->
+                                                                doesElementBelongToGivenFile(
+                                                                        thrownType, NEW_PREFIX));
+                            }
+                            return false;
+                        }
+
+                        @Override
+                        public void scan(CtElement element) {
+                            if (element != null
+                                    && !element.isImplicit()
+                                    && element.getPosition().isValidPosition()) {
+                                String pathString = element.getPath().toString();
+                                // Check if element is a child of an element corresponding to new
+                                // path
+                                if (insertedLines.stream()
+                                        .anyMatch(
+                                                line ->
+                                                        !(line.equals(pathString))
+                                                                && pathString.startsWith(line))) {
+                                    return;
+                                }
+
+                                // Check if any of the path in metadata file match in patched
+                                // program
+                                if (insertedLines.contains(pathString)
+                                        || isSetCollectionModified(element)) {
+                                    assertTrue(
+                                            doesElementBelongToGivenFile(element, NEW_PREFIX),
+                                            "Element should originate from new file but does not");
+                                }
+                                // Case when there is no entry of the path of the element in the
+                                // metadata file
+                                else {
+                                    assertTrue(
+                                            doesElementBelongToGivenFile(element, PREV_PREFIX),
+                                            "Element should originate from prev file but does not");
+                                }
+                            }
+                            super.scan(element);
+                        }
+                    };
+            scanner.scan(patchedCtModel.getRootPackage());
```

Comment:
This is a massive amount of code to put into a single method. I would avoid using anonymous classes for anything but very, very tiny things. Really, in Java 8 I'd only use lambdas.

Reply:
I will try to refactor this. This will be a good exercise for me after today's meeting.